src/osm-stop/MultiOsmStopSpell.sol

```diff
@@ -0,0 +1,122 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.8.16;
+
+import {DssEmergencySpell} from "../DssEmergencySpell.sol";
+
+interface IlkRegistryLike {
+    function count() external view returns (uint256);
+    function list() external view returns (bytes32[] memory);
+    function list(uint256 start, uint256 end) external view returns (bytes32[] memory);
+}
+
+interface OsmMomLike {
+    function stop(bytes32 ilk) external;
+    function osms(bytes32 ilk) external view returns (address);
+}
+
+interface OsmLike {
+    function stopped() external view returns (uint256);
+    function wards(address who) external view returns (uint256);
+}
+
+contract MultiOsmStopSpell is DssEmergencySpell {
+    string public constant override description = "Emergency Spell | Multi OSM Stop";
+
+    IlkRegistryLike public immutable ilkReg = IlkRegistryLike(_log.getAddress("ILK_REGISTRY"));
+    OsmMomLike public immutable osmMom = OsmMomLike(_log.getAddress("OSM_MOM"));
+
+    event Stop(bytes32 indexed ilk, address indexed osm);
+
+    /**
+     * @notice Stops, when possible, all OSMs that can be found through the ilk registry.
+     */
+    function _emeregencyActions() internal override {
+        bytes32[] memory ilks = ilkReg.list();
+        _doStop(ilks);
+    }
+
+    /**
+     * @notice Stops all OSMs in the batch.
+     * @dev This is an escape hatch to prevent this spell from being blocked in case it would hit the block gas limit.
+     *      In case `end` is greater than the ilk registry length, the iteration will be automatically capped.
+     * @param start The index to start the iteration (inclusive).
+     * @param end The index to stop the iteration (inclusive).
+     */
+    function stopBatch(uint256 start, uint256 end) external {
+        uint256 maxEnd = ilkReg.count() - 1;
+        bytes32[] memory ilks = ilkReg.list(start, end < maxEnd ? end : maxEnd);
+        _doStop(ilks);
+    }
+
+    /**
+     * @notice Stops, when possible, all OSMs that can be found from the `ilks` list.
+     * @param ilks The list of ilks to consider.
+     */
+    function _doStop(bytes32[] memory ilks) internal {
+        for (uint256 i = 0; i < ilks.length; i++) {
+            address osm = osmMom.osms(ilks[i]);
+
+            if (osm == address(0)) continue;
+
+            try OsmLike(osm).wards(address(osmMom)) returns (uint256 ward) {
+                // Ignore Osm instances that have not relied on OsmMom.
+                if (ward == 0) continue;
+            } catch Error(string memory reason) {
+                // If the reason is empty, it means the contract is most likely not an OSM instance.
+                require(bytes(reason).length == 0, reason);
```

Comment:
If for some reason one of the contracts revert with a reason (i.e: wrong auth/ward setup), this spell will revert and fail to stop all of them. Since this is to be used in an emergency situation it might be better to stop all that are possible and emit events for failures to be checked afterwards.

Reply:
The strategy is pretty much the same discussed above: we try to anticipate whatever reason the call would have to fail and skip that ilk if it does.
At this point, the only possibility would be that it fails because the registry is wrong, and the contract is not an OSM.